### PR TITLE
W5 - nhatnguyen - Thread Handle CountDown Timer

### PR DESCRIPTION
**Pull request by at-nhatnguyen**
#### Related issues (Optional)
Task #66 

#### Description (Required)
- [x] Thread Down Load Image
- [x] CountDown add and remove RecyclerView

### CODE REVIEW

File: app/src/at-nhatnguyen/java/vn/asiantech/internship/ui/MainActivity.java [hunk 1]
```diff
@@ -8,12 +8,13 @@
...
 import vn.asiantech.internship.R;
 import vn.asiantech.internship.ui.calculator.CalculatorActivity;
-import vn.asiantech.internship.ui.imageview.ImageActivity;
-import vn.asiantech.internship.ui.login.LoginActivity;
 import vn.asiantech.internship.ui.drawerlayout.DrawerActivity;
+import vn.asiantech.internship.ui.imageview.ImageActivity;
 import vn.asiantech.internship.ui.login.LoginActivity;
 import vn.asiantech.internship.ui.recyclerview.CommentActivity;
 import vn.asiantech.internship.ui.savedata.SaveDataActivity;
+import vn.asiantech.internship.ui.thread.LoadImageActivity;
```
Comment (at-dev-support-01) [inline app/src/at-nhatnguyen/java/vn/asiantech/internship/ui/MainActivity.java]: WARNING: [UnusedImports] Avoid unused imports such as 'vn.asiantech.internship.ui.thread.LoadImageActivity'
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/imports.html#UnusedImports

File: app/src/at-nhatnguyen/java/vn/asiantech/internship/ui/thread/fragment/OneFragment.java [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+                        handler.post(new Runnable() {
+                            @Override
+                            public void run() {
+                                mProgressBar.setProgress(mPercent);
+                            }
+                        });
+                        mBitmap = BitmapFactory.decodeByteArray(outputStream.toByteArray(), 0, outputStream.toByteArray().length);
+                    }
+                } catch (IOException e) {
+                    e.printStackTrace();
```
Comment (at-dev-support-01) [inline app/src/at-nhatnguyen/java/vn/asiantech/internship/ui/thread/fragment/OneFragment.java]: WARNING: [AvoidPrintStackTrace] Avoid printStackTrace(); use a logger call instead.
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/logging-java.html#AvoidPrintStackTrace

File: app/src/at-nhatnguyen/java/vn/asiantech/internship/ui/thread/fragment/OneFragment.java [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+                } catch (InterruptedException e) {
+                    e.printStackTrace();
```
Comment (at-dev-support-01) [inline app/src/at-nhatnguyen/java/vn/asiantech/internship/ui/thread/fragment/OneFragment.java]: WARNING: [AvoidPrintStackTrace] Avoid printStackTrace(); use a logger call instead.
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/logging-java.html#AvoidPrintStackTrace

File: app/src/at-nhatnguyen/java/vn/asiantech/internship/ui/thread/fragment/OneFragment.java [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+                } finally {
+                    try {
+                        if (connection != null)
```
Comment (at-dev-support-01) [inline app/src/at-nhatnguyen/java/vn/asiantech/internship/ui/thread/fragment/OneFragment.java]: WARNING: [IfStmtsMustUseBraces] Avoid using if statements without curly braces
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/braces.html#IfStmtsMustUseBraces

File: app/src/at-nhatnguyen/java/vn/asiantech/internship/ui/thread/fragment/OneFragment.java [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+                            connection.disconnect();
+                        if (inputStream != null)
```
Comment (at-dev-support-01) [inline app/src/at-nhatnguyen/java/vn/asiantech/internship/ui/thread/fragment/OneFragment.java]: WARNING: [IfStmtsMustUseBraces] Avoid using if statements without curly braces
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/braces.html#IfStmtsMustUseBraces

File: app/src/at-nhatnguyen/java/vn/asiantech/internship/ui/thread/fragment/OneFragment.java [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+                            inputStream.close();
+                    } catch (Exception e) {
+                        e.printStackTrace();
```
Comment (at-dev-support-01) [inline app/src/at-nhatnguyen/java/vn/asiantech/internship/ui/thread/fragment/OneFragment.java]: WARNING: [AvoidPrintStackTrace] Avoid printStackTrace(); use a logger call instead.
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/logging-java.html#AvoidPrintStackTrace

**Comment by at-nhatnguyen** [comment]
@at-kietvo please review my task, thank you

File: app/src/at-nhatnguyen/java/vn/asiantech/internship/ui/thread/CountDownActivity.java [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+    }
+
+    private void initAdapter() {
+        mCountDownAdapter = new CountDownAdapter(mNumber);
+        mRecycleViewCountDown.setLayoutManager(new LinearLayoutManager(this));
+        mRecycleViewCountDown.setAdapter(mCountDownAdapter);
+    }
+
+    private void initRun() {
+        CountDownTimer countDownTimer = new CountDownTimer(180000, 1000) {
```
Comment (at-kietvo) [inline app/src/at-nhatnguyen/java/vn/asiantech/internship/ui/thread/CountDownActivity.java]: Pls define the number to `constant` variable

Comment (at-kietvo) [reply app/src/at-nhatnguyen/java/vn/asiantech/internship/ui/thread/CountDownActivity.java]: and reuse it

File: app/src/at-nhatnguyen/java/vn/asiantech/internship/ui/thread/CountDownActivity.java [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+                mCountDownAdapter.notifyDataSetChanged();
+            }
+
+            @Override
+            public void onFinish() {
+                //NO-OP
+            }
+        };
+
+        countDownTimer.start();
```
Comment (at-kietvo) [inline app/src/at-nhatnguyen/java/vn/asiantech/internship/ui/thread/CountDownActivity.java]: Pls define count down timer variable in global and `cancel()` it after use in `onDestroy()`

File: app/src/at-nhatnguyen/java/vn/asiantech/internship/ui/thread/fragment/OneFragment.java [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+    }
+
+    private void downloadPhoto() {
+        new Thread(new Runnable() {
+            @Override
+            public void run() {
+                HttpURLConnection connection = null;
+                InputStream inputStream = null;
+                try {
+                    Thread.sleep(1000);
```
Comment (at-kietvo) [inline app/src/at-nhatnguyen/java/vn/asiantech/internship/ui/thread/fragment/OneFragment.java]: Why need sleep 1s ?

File: app/src/at-nhatnguyen/java/vn/asiantech/internship/ui/thread/fragment/OneFragment.java [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+                    connection = (HttpURLConnection) new URL(LINK_PICTURE).openConnection();
+                    connection.setDoInput(true);
+                    connection.connect();
+                    int fileLength = connection.getContentLength();
+                    inputStream = connection.getInputStream();
+                    ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+                    byte data[] = new byte[128];
+                    long total = 0;
+                    int count;
+                    while ((count = inputStream.read(data)) != -1) {
```
Comment (at-kietvo) [inline app/src/at-nhatnguyen/java/vn/asiantech/internship/ui/thread/fragment/OneFragment.java]: You are limiting the part of the total for reading stream, pls remove `data`, we can use `inputStream.read()` instead.
Also remove variable of `data[]` and change condition of `while`

File: app/src/at-nhatnguyen/java/vn/asiantech/internship/ui/thread/fragment/OneFragment.java [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+                        total += count;
+                        mPercent = (int) (total * 100 / fileLength);
+                        outputStream.write(data, 0, count);
+                        handler.post(new Runnable() {
+                            @Override
+                            public void run() {
+                                mProgressBar.setProgress(mPercent);
+                            }
+                        });
+                        mBitmap = BitmapFactory.decodeByteArray(outputStream.toByteArray(), 0, outputStream.toByteArray().length);
```
Comment (at-kietvo) [inline app/src/at-nhatnguyen/java/vn/asiantech/internship/ui/thread/fragment/OneFragment.java]: Pls move this way out site `while` for performance

File: app/src/at-nhatnguyen/java/vn/asiantech/internship/ui/thread/fragment/TwoFragment.java [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+        View view = inflater.inflate(R.layout.fragment_two, container, false);
+        initViews(view);
+        return view;
+    }
+
+    private void initViews(View view) {
+        mImgView = view.findViewById(R.id.imgPhoto);
+    }
+
+    public void setMImgView(Bitmap bitmap) {
```
Comment (at-kietvo) [inline app/src/at-nhatnguyen/java/vn/asiantech/internship/ui/thread/fragment/TwoFragment.java]: change method name

File: app/src/at-nhatnguyen/java/vn/asiantech/internship/ui/thread/fragment/OneFragment.java [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+                            inputStream.close();
+                        }
+                    } catch (Exception e) {
+                        Log.d("", e.getMessage());
+                    }
+                }
+                getActivity().runOnUiThread(new Runnable() {
+                    @Override
+                    public void run() {
+                        ((LoadImageActivity) getActivity()).getBitmap(mBitmap);
```
Comment (at-kietvo) [inline app/src/at-nhatnguyen/java/vn/asiantech/internship/ui/thread/fragment/OneFragment.java]: check `null`

File: app/src/at-nhatnguyen/java/vn/asiantech/internship/ui/thread/LoadImageActivity.java [hunk 14]
```diff
@@ -0,0 +1,47 @@
+package vn.asiantech.internship.ui.thread;
+
+import android.graphics.Bitmap;
+import android.os.Bundle;
+import android.support.design.widget.TabLayout;
+import android.support.v4.view.ViewPager;
+import android.support.v7.app.AppCompatActivity;
+import android.util.Log;
```
Comment (at-dev-support-01) [inline app/src/at-nhatnguyen/java/vn/asiantech/internship/ui/thread/LoadImageActivity.java]: WARNING: [UnusedImports] Avoid unused imports such as 'android.util.Log'
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/imports.html#UnusedImports

File: app/src/at-nhatnguyen/java/vn/asiantech/internship/ui/thread/fragment/TwoFragment.java [hunk 15]
```diff
@@ -0,0 +1,44 @@
+package vn.asiantech.internship.ui.thread.fragment;
+
+import android.graphics.Bitmap;
+import android.os.Bundle;
+import android.support.annotation.NonNull;
+import android.support.v4.app.Fragment;
+import android.util.Log;
```
Comment (at-dev-support-01) [inline app/src/at-nhatnguyen/java/vn/asiantech/internship/ui/thread/fragment/TwoFragment.java]: WARNING: [UnusedImports] Avoid unused imports such as 'android.util.Log'
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/imports.html#UnusedImports

**Comment by at-nhatnguyen** [comment]
@at-kietvo please review my task, thank you

**Comment by at-kietvo** [comment]
@at-nhatnguyen : Pls check conflict.

**Comment by at-nhatnguyen** [comment]
@at-kietvo please review my task, thanks

**Pull request merged**